torch/utils/flop_counter.py

```diff
@@ -0,0 +1,306 @@
+    """
+    batch_size = x_shape[0]
+    conv_shape = (x_shape if transposed else out_shape)[2:]
+    c_out, c_in, *dims = w_shape
+
+    # NB(chilli): I don't think this properly accounts for padding :think:
+    # NB(chilli): Should be 2 * c_in - 1 technically for FLOPs.
+    flop = batch_size * prod(conv_shape) * c_out * prod(dims) * 2 * c_in
+    return flop
+
+def conv_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    """
+    Count flops for convolution.
+    """
+    x_shape, w_shape = input_shapes[:2]
+    out_shape = output_shapes[0]
+    transposed = input_shapes[6]
+
+    return conv_flop_count(x_shape, w_shape, out_shape, transposed=transposed)
+
+def transpose_shape(shape):
+    return [shape[1], shape[0]] + list(shape[2:])
+
+def conv_backward_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    grad_out_shape, x_shape, w_shape = input_shapes[:3]
+    output_mask = input_shapes[-1]
+    fwd_transposed = input_shapes[7]
+    flop_count = 0
+
+    if output_mask[0]:
+        grad_input_shape = get_shape(output_shapes[0])
+        flop_count += conv_flop_count(grad_out_shape, w_shape, grad_input_shape, not fwd_transposed)
+    if output_mask[1]:
+        grad_weight_shape = get_shape(output_shapes[1])
+        flop_count += conv_flop_count(transpose_shape(x_shape), grad_out_shape, grad_weight_shape, fwd_transposed)
+
+    return flop_count
+
+
+flop_mapping = {
+    aten.mm: mm_flop,
+    aten.addmm: addmm_flop,
+    aten.bmm: bmm_flop,
+    aten.baddbmm: baddbmm_flop,
+    aten.convolution: conv_flop,
+    aten._convolution: conv_flop,
+    aten.convolution_backward: conv_backward_flop,
+}
+
+def normalize_tuple(x):
+    if not isinstance(x, tuple):
+        return (x,)
+    return x
+
+
+# Thanks BingChat!
+def convert_num_to_suffix_str(number):
+    # Define the suffixes for different orders of magnitude
+    suffixes = ["", "K", "M", "B", "T"]
+    # Find the index of the appropriate suffix based on the number of digits
+    index = max(0, min(len(suffixes) - 1, (len(str(number)) - 1) // 3))
+    # Divide the number by 1000^index and format it to two decimal places
+    value = "{:.2f}".format(number / (1000 ** index))
+    # Return the value and the suffix as a string
+    return value + suffixes[index]
+
+class FlopCounterMode(TorchDispatchMode):
+    """
+    ``FlopCounterMode`` is a context manager that counts the number of
+    flops within its context. It does this using a ``TorchDispatchMode``.
+
+    It also supports hierarchical output by passing a module to FlopCounterMode on construction.
+
+    Example usage
+
+    .. code-block:: python
+
+        mod = ...
+        flop_counter = FlopCounterMode(mod)
+        with flop_counter:
+            mod.sum().backward()
+
+    """
+    def __init__(self, mod: torch.nn.Module = None, depth: int = 2, display: bool = True, custom_mapping: Dict[Any, Any] = None):
+        self.flop_counts: Dict[Any, Any] = defaultdict(lambda: defaultdict(int))
```

Comment:
Users want to access this, either for a custom display format, or for aggregating it across a dataset (e.g. https://github.com/facebookresearch/detectron2/blob/38af375052d3ae7331141bc1a22cfa2713b02987/tools/analyze_model.py#L66-L68)

So it's probably better to make it a proper API with documentation.